AK/BinarySearch.h

```diff
@@ -32,14 +32,16 @@
 
 namespace AK {
 
-template<typename T>
-constexpr int integral_compare(const typename RemoveConst<T>::Type& a, const typename RemoveConst<T>::Type& b)
-{
-    return a - b;
-}
+struct IntegralComperator {
+    constexpr auto operator()(auto& lhs, auto& rhs) { return lhs - rhs; }
+};
 
-template<typename T, typename Compare>
-constexpr T* binary_search(Span<T> haystack, const typename RemoveConst<T>::Type& needle, Compare compare = integral_compare, size_t* nearby_index = nullptr)
+template<typename Container, typename Needle, typename Comperator = IntegralComperator>
+constexpr auto binary_search(
```

Comment:
Does `decltype(auto)` work here?
it'd help the return type issue if the compiler doesn't get mad at it

Reply:
I do get an error in that case:

~~~
inconsistent deduction for auto return type: 'std::nullptr_t' and then 'const unsigned int*'
~~~

I also tried leaving out the trailing return type entirely (just `auto` at the beginning) and get the same error.

Reply:
oh it's also returning `nullptr` at some point...:sad: